In [1]:
import numpy as np
import keras
from keras.datasets import fashion_mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.optimizers import SGD, Adam
from keras.utils import to_categorical
from sklearn.metrics import accuracy_score

Using TensorFlow backend.
C:\Users\ataga\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\Users\ataga\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\Users\ataga\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\Users\ataga\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py

In [2]:
np.random.seed(1)

In [3]:
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

In [4]:
# Normalize the images.
x_train = (x_train / 255) - 0.5
x_test = (x_test / 255) - 0.5

# Flatten the images.
x_train = x_train.reshape((-1, 784))
x_test = x_test.reshape((-1, 784))

print(x_train.shape) # (60000, 784)
print(x_test.shape)  # (10000, 784)

(60000, 784)
(10000, 784)


In [5]:
# Build the model.
# model = Sequential([
#   Dense(128, activation='relu', input_shape=(784,)),
#   Dense(128, activation='relu'),
#   Dense(10, activation='softmax'),
# ])

N_EPOCHS = 100
BATCH_SIZE = 128
N_CLASSES = 10
OPTIMISER = Adam()
N_HIDDEN = 256
DROPOUT = 0.2

model = Sequential()
model.add(Dense(N_HIDDEN, activation='relu', input_shape=(784,)))
model.add(Dropout(DROPOUT))
model.add(Dense(N_HIDDEN, activation='relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(N_CLASSES, activation='softmax'))
model.summary()

Instructions for updating:
Colocations handled automatically by placer.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 256)               200960    
_________________________________________________________________
dropout_1 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 256)               65792     
_________________________________________________________________
dropout_2 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                2570      
Total params: 269,322
Trainable params: 269,322
Non-trainable par

In [6]:
# Compile the model.
model.compile(
  optimizer=OPTIMISER,
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

In [7]:
# Train the model.
model.fit(
    x_train,
    to_categorical(y_train),
    epochs=N_EPOCHS,
    batch_size=BATCH_SIZE,
    validation_split=0.2
)

Instructions for updating:
Use tf.cast instead.
Train on 48000 samples, validate on 12000 samples
Epoch 1/100
48000/48000 [==============================] - 2s 40us/step - loss: 0.5717 - acc: 0.7928 - val_loss: 0.4152 - val_acc: 0.8484
Epoch 2/100
48000/48000 [==============================] - 2s 31us/step - loss: 0.4132 - acc: 0.8490 - val_loss: 0.3725 - val_acc: 0.8619
Epoch 3/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.3774 - acc: 0.8608 - val_loss: 0.3552 - val_acc: 0.8719
Epoch 4/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.3512 - acc: 0.8713 - val_loss: 0.3419 - val_acc: 0.8753
Epoch 5/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.3352 - acc: 0.8761 - val_loss: 0.3375 - val_acc: 0.8792
Epoch 6/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.3174 - acc: 0.8823 - val_loss: 0.3398 - val_acc: 0.8762
Epoch 7/100
48000/48000 [==============================] - 2s 33us/step - 

Epoch 58/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.1180 - acc: 0.9551 - val_loss: 0.3809 - val_acc: 0.9008
Epoch 59/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.1181 - acc: 0.9567 - val_loss: 0.3789 - val_acc: 0.8982
Epoch 60/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.1161 - acc: 0.9560 - val_loss: 0.3942 - val_acc: 0.8955
Epoch 61/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.1127 - acc: 0.9572 - val_loss: 0.3925 - val_acc: 0.8936
Epoch 62/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.1104 - acc: 0.9576 - val_loss: 0.4010 - val_acc: 0.8964
Epoch 63/100
48000/48000 [==============================] - 1s 30us/step - loss: 0.1143 - acc: 0.9553 - val_loss: 0.3871 - val_acc: 0.8961
Epoch 64/100
48000/48000 [==============================] - 1s 31us/step - loss: 0.1085 - acc: 0.9588 - val_loss: 0.3837 - val_acc: 0.8990
Epoch 65/100
48000/48000 [=

In [8]:
# Evaluate the model.
model.evaluate(
  x_test,
  to_categorical(y_test)
)

10000/10000 [==============================] - 0s 18us/step


[0.47610057914853093, 0.891]

In [9]:
# Save the model to disk.
model.save_weights('nn_model_hw_2.h5')

In [10]:
# Load the model from disk later using:
# model.load_weights('nn_model_hw_2.h5')

In [11]:
# Predict on the first 5 test images.
predictions = model.predict(x_test)

In [12]:
test_preds = np.argmax(predictions, axis=1)

In [13]:
accuracy_score(y_test, test_preds)

0.891

Я увеличила число входных неронов и нейронов скрытого слоя до 256, добавила прореживание при помощи метода Dropout, увеличила размер батча до 128, при обучении использовала валидацию на отложенной выборке, и в итоге получиал 89% точность на тесте.